File: packages/ormai-ts/src/adapters/typeorm/introspection.ts
/**
 * TypeORM schema introspection.
 */

import type {
  FieldMetadata,
  FieldType,
  ModelMetadata,
  RelationMetadata,
  RelationType,
  SchemaMetadata,
} from '../../core/types.js';

/**
 * TypeORM column metadata.
 */
export interface TypeORMColumnMetadata {
  propertyName: string;
  databaseName: string;
  type: string | Function;
  isPrimary: boolean;
  isNullable: boolean;
  default?: unknown;
}

/**
 * TypeORM relation metadata.
 */
export interface TypeORMRelationMetadata {
  propertyName: string;
  relationType: 'one-to-one' | 'one-to-many' | 'many-to-one' | 'many-to-many';
  inverseSidePropertyPath?: string;
  type: string | Function;
}

/**
 * TypeORM entity metadata.
 */
export interface TypeORMEntityMetadata {
  name: string;
  tableName: string;
  columns: TypeORMColumnMetadata[];
  relations: TypeORMRelationMetadata[];
  primaryColumns: TypeORMColumnMetadata[];
}

/**
 * TypeORM data source for introspection.
 */
export interface TypeORMDataSource {
  entityMetadatas: TypeORMEntityMetadata[];
  getMetadata: (entity: string | Function) => TypeORMEntityMetadata;
}

/**
 * TypeORM introspector for converting TypeORM metadata to OrmAI metadata.
 */
export class TypeORMIntrospector {
  private readonly dataSource: TypeORMDataSource;
  private readonly modelFilter?: string[];

  constructor(dataSource: TypeORMDataSource, models?: string[]) {
    this.dataSource = dataSource;
    this.modelFilter = models;
  }

  /**
   * Convert TypeORM metadata to OrmAI SchemaMetadata.
   */
  introspect(): SchemaMetadata {
    const models: Record<string, ModelMetadata> = {};

    for (const entityMetadata of this.dataSource.entityMetadatas) {
      const modelName = entityMetadata.name;

      // Filter models if specified
      if (this.modelFilter && !this.modelFilter.includes(modelName)) {
        continue;
      }

      const fields: Record<string, FieldMetadata> = {};
      const primaryKeys: string[] = [];

      // Process columns
      for (const column of entityMetadata.columns) {
        fields[column.propertyName] = {
          name: column.propertyName,
          fieldType: this.mapColumnType(column.type),
          nullable: column.isNullable,
          primaryKey: column.isPrimary,
          default: column.default,
        };

        if (column.isPrimary) {
          primaryKeys.push(column.propertyName);
        }
      }

      // Process relations
      const relations: Record<string, RelationMetadata> = {};
      for (const relation of entityMetadata.relations) {
        const targetName = typeof relation.type === 'function'
          ? (relation.type as Function).name
          : String(relation.type);

        relations[relation.propertyName] = {
          name: relation.propertyName,
          relationType: this.mapRelationType(relation.relationType),
          targetModel: targetName,
          foreignKey: relation.inverseSidePropertyPath,
        };
      }

      const primaryKey = primaryKeys.length > 0 ? primaryKeys[0] : 'id';

      models[modelName] = {
        name: modelName,
        tableName: entityMetadata.tableName,
        fields,
        relations,
        primaryKey,
        primaryKeys: primaryKeys.length > 1 ? primaryKeys : undefined,
      };
    }

    return { models };
  }

  /**
   * Map TypeORM column type to OrmAI field type.
   */
  private mapColumnType(type: string | Function): FieldType {
    const typeStr = typeof type === 'function' ? type.name.toLowerCase() : String(type).toLowerCase();

    // String types
    if (typeStr === 'string' || typeStr.includes('varchar') || typeStr.includes('char') || typeStr.includes('text')) {
      return 'string';
    }

    // Integer types
    if (typeStr === 'number' || typeStr.includes('int') || typeStr === 'integer') {
      return 'integer';
    }

    // Float types
    if (typeStr.includes('float') || typeStr.includes('double') || typeStr.includes('decimal') || typeStr.includes('numeric')) {
      return 'float';
    }

    // Boolean
    if (typeStr === 'boolean' || typeStr === 'bool') {
      return 'boolean';
    }

    // Date/Time
    if (typeStr === 'date') {
      return 'date';
    }
    if (typeStr.includes('timestamp') || typeStr.includes('datetime')) {
      return 'datetime';
    }
    if (typeStr.includes('time')) {
      return 'time';
    }

    // JSON
    if (typeStr === 'json' || typeStr === 'jsonb' || typeStr === 'object') {
      return 'json';
    }

    // UUID
    if (typeStr === 'uuid') {
      return 'uuid';
    }

    // Binary - map to string since bytes is not a valid type
    if (typeStr.includes('blob') || typeStr.includes('binary') || typeStr === 'buffer') {
      return 'string';
    }

    // Default
    return 'string';
  }

  /**
   * Map TypeORM relation type to OrmAI relation type.
   */
  private mapRelationType(type: string): RelationType {
    switch (type) {
      case 'one-to-one':
        return 'one_to_one';
      case 'one-to-many':
        return 'one_to_many';
      case 'many-to-one':
        return 'many_to_one';
      case 'many-to-many':
        return 'many_to_many';
      default:
        return 'many_to_one';
    }
  }
}

/**
 * Create a TypeORM introspector.
 */
export function createTypeORMIntrospector(
  dataSource: TypeORMDataSource,
  models?: string[]
): TypeORMIntrospector {
  return new TypeORMIntrospector(dataSource, models);
}

File: src/ormai/adapters/tortoise/introspection.py
"""
Tortoise ORM schema introspection.

Extracts schema metadata from Tortoise model definitions.
"""

from __future__ import annotations

from typing import Any

from ormai.core.types import (
    FieldMetadata,
    ModelMetadata,
    RelationMetadata,
    RelationType,
    SchemaMetadata,
)


class TortoiseIntrospector:
    """
    Introspects Tortoise ORM models to build schema metadata.
    """

    def __init__(self, models: list[type]) -> None:
        """
        Initialize the introspector.

        Args:
            models: List of Tortoise model classes
        """
        self.models = models

    def introspect(self) -> SchemaMetadata:
        """
        Introspect all registered models and build schema metadata.
        """
        model_metadata: dict[str, ModelMetadata] = {}

        for model in self.models:
            model_meta = self._introspect_model(model)
            model_metadata[model_meta.name] = model_meta

        return SchemaMetadata(models=model_metadata)

    def _introspect_model(self, model: type) -> ModelMetadata:
        """Introspect a single Tortoise model."""
        from tortoise.fields.relational import (
            BackwardFKRelation,
            BackwardOneToOneRelation,
        )
        from tortoise.models import Model

        if not issubclass(model, Model):
            raise ValueError(f"{model} is not a Tortoise Model")

        meta = model._meta

        # Get fields
        fields_meta: dict[str, FieldMetadata] = {}
        for field_name, field_obj in meta.fields_map.items():
            # Skip reverse relations
            if isinstance(field_obj, (BackwardFKRelation, BackwardOneToOneRelation)):
                continue

            field_meta = self._introspect_field(field_name, field_obj)
            if field_meta:
                fields_meta[field_name] = field_meta

        # Get relations
        relations_meta: dict[str, RelationMetadata] = {}
        for field_name, field_obj in meta.fields_map.items():
            rel_meta = self._introspect_relation(field_name, field_obj, model)
            if rel_meta:
                relations_meta[field_name] = rel_meta

        # Find primary key
        pk_field = meta.pk_attr or "id"

        return ModelMetadata(
            name=model.__name__,
            table_name=meta.db_table or model.__name__.lower(),
            fields=fields_meta,
            relations=relations_meta,
            primary_key=pk_field,
        )

    def _introspect_field(self, name: str, field: Any) -> FieldMetadata | None:
        """Introspect a single field."""
        from tortoise import fields as tortoise_fields
        from tortoise.fields.relational import (
            BackwardFKRelation,
            BackwardOneToOneRelation,
            ForeignKeyFieldInstance,
            ManyToManyFieldInstance,
            OneToOneFieldInstance,
        )

        # Map Tortoise field types to our type system
        type_map: dict[type, str] = {
            tortoise_fields.IntField: "integer",
            tortoise_fields.BigIntField: "bigint",
            tortoise_fields.SmallIntField: "smallint",
            tortoise_fields.CharField: "string",
            tortoise_fields.TextField: "text",
            tortoise_fields.BooleanField: "boolean",
            tortoise_fields.DecimalField: "decimal",
            tortoise_fields.FloatField: "float",
            tortoise_fields.DateField: "date",
            tortoise_fields.DatetimeField: "datetime",
            tortoise_fields.TimeField: "time",
            tortoise_fields.UUIDField: "uuid",
            tortoise_fields.JSONField: "json",
            tortoise_fields.BinaryField: "binary",
        }

        # Handle relation fields separately
        if isinstance(field, (
            ForeignKeyFieldInstance,
            OneToOneFieldInstance,
            ManyToManyFieldInstance,
            BackwardFKRelation,
            BackwardOneToOneRelation,
        )):
            return None

        # Get the field type
        field_type = "unknown"
        for field_class, type_name in type_map.items():
            if isinstance(field, field_class):
                field_type = type_name
                break

        # Check if it's a primary key
        is_pk = getattr(field, "primary_key", False) or getattr(field, "pk", False)

        # Check if nullable
        nullable = getattr(field, "null", False)

        return FieldMetadata(
            name=name,
            field_type=field_type,
            nullable=nullable,
            primary_key=is_pk,
        )

    def _introspect_relation(
        self, name: str, field: Any, model: type  # noqa: ARG002
    ) -> RelationMetadata | None:
        """Introspect a relation field."""
        from tortoise.fields.relational import (
            BackwardFKRelation,
            BackwardOneToOneRelation,
            ForeignKeyFieldInstance,
            ManyToManyFieldInstance,
            OneToOneFieldInstance,
        )

        if isinstance(field, ForeignKeyFieldInstance):
            related_model = field.related_model
            return RelationMetadata(
                name=name,
                relation_type=RelationType.MANY_TO_ONE,
                target_model=related_model.__name__ if related_model else "Unknown",
                foreign_key=f"{name}_id",
            )

        if isinstance(field, OneToOneFieldInstance):
            related_model = field.related_model
            return RelationMetadata(
                name=name,
                relation_type=RelationType.ONE_TO_ONE,
                target_model=related_model.__name__ if related_model else "Unknown",
                foreign_key=f"{name}_id",
            )

        if isinstance(field, ManyToManyFieldInstance):
            related_model = field.related_model
            return RelationMetadata(
                name=name,
                relation_type=RelationType.MANY_TO_MANY,
                target_model=related_model.__name__ if related_model else "Unknown",
            )

        if isinstance(field, BackwardFKRelation):
            related_model = field.related_model
            return RelationMetadata(
                name=name,
                relation_type=RelationType.ONE_TO_MANY,
                target_model=related_model.__name__ if related_model else "Unknown",
            )

        if isinstance(field, BackwardOneToOneRelation):
            related_model = field.related_model
            return RelationMetadata(
                name=name,
                relation_type=RelationType.ONE_TO_ONE,
                target_model=related_model.__name__ if related_model else "Unknown",
            )

        return None
